### PR TITLE
Resolve #56: Use hkd

### LINKED ISSUE #56: Place and name of higher Functor
So we don't forget to settle on https://github.com/ekmett/adjunctions/pull/49#issuecomment-382416923

Based on todays comments by Edward, `Functor1` is not a good name!

---

https://github.com/ekmett/codex/blob/d4eeebf55a8c2bc9b858cdc0ddccb2679e03091f/lib/hkd/src/Data/HKD.hs#L61

```
21:48  edwardk Functor1, etc. doesn't work because I want FFunctor, FFoldable, FTraversable..
21:49  edwardk and the latter become Foldable1 and Traversable1 under that convention!
```

### PR DESCRIPTION
**Pull request by phadej**
The only "drawback" is that we only support GHC-7.6 and above, also for `adjunctons`. But who can live without `PolyKinds` nowadays?
(The real reason is that we cannot export `type (~>)` on older GHC).

ping @ekmett @RyanGlScott 

`hkd` defines natural transformations stuff, which is overlap with `natural-transformation` package. `Nat` in `hkd` doesn't have `$$` field accessor. I didn't change that.

I *might* myself need `hkd` soon too, so having it uploaded to Hackage would be great. I can be a co-maintainer.

If this direction is approved, I'd write haddocks for `hkd` before upload.

### CODE REVIEW

**Comment by RyanGlScott** [comment]
This seems plausible to me, although I'd prefer if @ekmett signed off on this before moving ahead.

File: hkd/hkd.cabal [hunk 1]
```diff
@@ -0,0 +1,61 @@
+cabal-version: 2.2
+name: hkd
+version: 0
+synopsis: "higher-kinded data"
+description:
+  "Higher-kinded data" utilities.
+homepage: https://github.com/ekmett/codex/tree/master/hkd#readme
```
Comment (quasicomputational) [inline hkd/hkd.cabal]: Presumably, this will change?

Comment (phadej) [reply hkd/hkd.cabal]: Yes. As implied, I didn't put too much work into human/documentation side of things yet.

**Comment by ekmett** [comment]
I’m planning on putting a new generation of distributive into that code as well using the approach Aaron Vargo came up with. Once that is done, I’d be happier pulling the trigger on an adjunctions refactor, as it’ll rip about half of this package out, by making distributive as powerful as representable is today, but more efficient.

**Comment by phadej** [comment]
`adjunctions` have been waiting for something to happen for 1,5 years (aaron’s stuff was merged in 2018).

I’d suggest this if there are no clear veto for this change, that we go forward with it now. And get current state of adjunctions released.

**Comment by ekmett** [comment]
The difference is I started _actually_ writing the code in question recently and don’t particularly want to thrash on two big releases in rapid succession.

**Comment by phadej** [comment]
@ekmett does that involve rewriting parts of `hkd` and/or including parts of it directly in `distributive` (or `adjunctions`?

Or to ask more directly, can `hkd` be uploaded to Hackage soon?

File: hkd/src/Data/HKD.hs [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+, flength
+, ftraverse_
+, ffor_
+-- * Traversable
+, FTraversable(..)
+, ffoldMapDefault
+, ffmapDefault
+, ffor
+-- * Utilities
+, Log(..)
```
Comment (phadej) [inline hkd/src/Data/HKD.hs]: In adjunctions, this was called `Logarithm`

File: hkd/src/Data/HKD.hs [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+, Tab(..)
+, indexLog
+, Element(..)
+, Nat(..)
+, A(..)
```
Comment (phadej) [inline hkd/src/Data/HKD.hs]: This type is called `Some` elsewhere

**Comment by phadej** [comment]
Switched to use [`hkd`](https://hackage.haskell.org/package/hkd)

**Comment by ekmett** [comment]
To be honest, after the feedback here, I got fairly frustrated and went off to do other things that I wouldn't have to fight my way uphill on.

**Comment by phadej** [comment]
@ekmett I apologize if my actions are perceived as negative, I only try to make `adjunctions` `master` be closer to the releasable state (e.g. if the release would be forced for external reasons).

If this is not the direction `adjunctions` will be moving, just close this pull request.

**Comment by ekmett** [comment]
That perhaps sounded way more negative than intended. I apologize.

**Comment by phadej** [comment]
Found another package in HKD space: https://hackage.haskell.org/package/czipwith-1.0.1.2/docs/Data-CZipWith.html

**Comment by phadej** [comment]
ping

**Comment by phadej** [comment]
As a one year anniversary of this PR, I'm dropping this initiative.

**Pull request closed without merging**